types/wlr_keyboard_shortcuts_inhibit_v1.c

```diff
@@ -0,0 +1,212 @@
+		&keyboard_shortcuts_inhibitor_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void keyboard_shortcuts_inhibitor_v1_destroy(
+		struct wlr_keyboard_shortcuts_inhibitor_v1 *inhibitor) {
+	if (!inhibitor) {
+		return;
+	}
+
+	wlr_signal_emit_safe(&inhibitor->events.destroy, inhibitor);
+
+	wl_resource_set_user_data(inhibitor->resource, NULL);
+	wl_list_remove(&inhibitor->manager_link);
+	wl_list_remove(&inhibitor->surface_destroy.link);
+	wl_list_remove(&inhibitor->seat_client_destroy.link);
+	free(inhibitor);
+}
+
+static void keyboard_shortcuts_inhibitor_v1_handle_resource_destroy(
+		struct wl_resource *resource) {
+	struct wlr_keyboard_shortcuts_inhibitor_v1 *inhibitor =
+		wlr_keyboard_shortcuts_inhibitor_v1_from_resource(resource);
+	keyboard_shortcuts_inhibitor_v1_destroy(inhibitor);
+}
+
+static void keyboard_shortcuts_inhibitor_handle_surface_destroy(
+		struct wl_listener *listener, void *data) {
+	struct wlr_keyboard_shortcuts_inhibitor_v1 *inhibitor =
+		wl_container_of(listener, inhibitor, surface_destroy);
+	keyboard_shortcuts_inhibitor_v1_destroy(inhibitor);
+}
+
+static void keyboard_shortcuts_inhibitor_handle_seat_client_destroy(
+		struct wl_listener *listener, void *data) {
+	struct wlr_keyboard_shortcuts_inhibitor_v1 *inhibitor =
+		wl_container_of(listener, inhibitor, seat_client_destroy);
+	keyboard_shortcuts_inhibitor_v1_destroy(inhibitor);
+}
+
+static void keyboard_shortcuts_inhibitor_v1_handle_destroy(
+		struct wl_client *client,
+		struct wl_resource *manager_resource) {
+	wl_resource_destroy(manager_resource);
+}
+
+static const struct zwp_keyboard_shortcuts_inhibitor_v1_interface
+keyboard_shortcuts_inhibitor_impl = {
+	.destroy = keyboard_shortcuts_inhibitor_v1_handle_destroy,
+};
+
+static void manager_handle_inhibit_shortcuts(struct wl_client *client,
+		struct wl_resource *resource, uint32_t id,
+		struct wl_resource *surface_resource,
+		struct wl_resource *seat_resource) {
+	struct wlr_surface *surface =
+		wlr_surface_from_resource(surface_resource);
+	struct wlr_seat_client *seat_client =
+		wlr_seat_client_from_resource(seat_resource);
+	struct wlr_keyboard_shortcuts_inhibit_manager_v1 *manager =
+		wlr_keyboard_shortcuts_inhibit_manager_v1_from_resource(resource);
+
+	struct wlr_keyboard_shortcuts_inhibitor_v1 *existing_inhibitor;
+	wl_list_for_each(existing_inhibitor, &manager->inhibitors, manager_link) {
+		if (existing_inhibitor->surface != surface ||
+				existing_inhibitor->seat_client != seat_client) {
+			continue;
+		}
+
+		wl_resource_post_error(resource,
+			ZWP_KEYBOARD_SHORTCUTS_INHIBIT_MANAGER_V1_ERROR_ALREADY_INHIBITED,
+			"this surface already has keyboard shortcuts "
+			"inhibited on this seat");
+		return;
+	}
+
+	struct wlr_keyboard_shortcuts_inhibitor_v1 *inhibitor =
+		calloc(1, sizeof(struct wlr_keyboard_shortcuts_inhibitor_v1));
+	if (!inhibitor) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	struct wl_resource *wl_resource = wl_resource_create(client,
```

Comment:
We have both `wl_resource` (the inhibitor) and `resource` (the manager). It would be nice to rename those to make them less confusing.

Reply:
Done.